File: Client-zhy/Assets/Lua/Mgr/UIMgr.lua
module("UIMgr", package.seeall)

--C#层的UIMgr
local mUIMgr = GameCore.UIMgr

local function DoInvoke(funcName)
    -- body
end

--以下是UI状态事件方法
local function OnCreate(uiID, frame)
    local uiData = AllUI.GetUIData(uiID)
    if uiData then
        if uiData.luaScript and uiData.luaScript.OnCreate then
            uiData.luaScript.OnCreate(frame)
        end
    else
        --没找到uidata，提示
        GameLog.LogError("UIMgr.OnCreate -> uiData is nil, uiID = %s", uiID)
    end
end

local function OnEnable(uiID, frame)
    local uiData = AllUI.GetUIData(uiID)
    if uiData then
        if uiData.luaScript and uiData.luaScript.OnEnable then
            uiData.luaScript.OnEnable(frame)
        end
    else
        GameLog.LogError("UIMgr.OnEnable -> uiData is nil, uiID = %s", uiID)
        
    end
end

local function OnDisable(uiID, frame)
    local uiData = AllUI.GetUIData(uiID)
    if uiData then
        if uiData.luaScript and uiData.luaScript.OnDisable then
            uiData.luaScript.OnDisable(frame)
        end
    else
        GameLog.LogError("UIMgr.OnDisable -> uiData is nil, uiID = %s", uiID)
    end
end

local function OnDestroy(uiID, frame)
    local uiData = AllUI.GetUIData(uiID)
    if uiData then
        if uiData.luaScript and uiData.luaScript.OnDestroy then
            uiData.luaScript.OnDestroy(frame)
        end
    else
        GameLog.LogError("UIMgr.OnDestroy -> uiData is nil, uiID = %s", uiID)
    end
end

--以下是NGUI事件方法
local function OnClick(uiID, eventID)
    local uiData = AllUI.GetUIData(uiID)
    if uiData then
        if uiData.luaScript and uiData.luaScript.OnClick then
            uiData.luaScript.OnClick(eventID)
        end
    else
        GameLog.LogError("UIMgr.OnClick -> uiData is nil, uiID = %s", uiID)
    end
end

local function OnPress(uiID, isPressed, eventID)
    local uiData = AllUI.GetUIData(uiID)
    if uiData then
        if uiData.luaScript and uiData.luaScript.OnPress then
            uiData.luaScript.OnPress(isPressed, eventID)
        end
    else
        
    end
end

local function OnSelect(uiID, selected, eventID)
end

local function OnDoubleClick(uiID, eventID)
end

local function OnDragStart(uiID, eventID)
end

local function OnDrag(uiID, delta, eventID)
end

local function OnDragEnd(uiID, eventID)
end

local function OnDragOver(uiID, obj, eventID)
end

local function OnDragOut(uiID, obj, eventID)
    
end

function Init()
    --可能还需要UIRoot作为根节点，放置新创建的UI
    mUIMgr.Init(OnCreate, OnEnable, OnDisable, OnDestroy,
            OnClick, OnPress, OnSelect,
            OnDoubleClick,
            OnDragStart, OnDrag, OnDragEnd, OnDragOver, OnDragOut)

    --

end

--UI打开的处理过程为异步后，需要再传递一个回调
--这个回调在lua层由统一的地方管理，业务逻辑层只需用uiID和一个lua方法即可注册
function ShowUI(uiData)
    if AllUI.CheckIsValid(uiData) then
        mUIMgr.ShowUI(uiData.uiID, uiData.uiName)
    else
        --uiData无效，提示
    end
end

function UnShowUI(uiData)
    if AllUI.CheckIsValid(uiData) then
        mUIMgr.UnShowUI(uiData.uiID)
    else
        --uiData无效，提示
    end
end

return UIMgr

File: Client-zhy/Assets/Code/Logic/GameCore/UIMgr.cs
using System.Collections;
using System.Collections.Generic;
using UnityEngine;
using LuaInterface;

//某些C#类，用于对lua提供服务，可能Lua层也有一份相同类名的.
//这时，就需要把C#类定义在该名空间下
namespace GameCore
{
    public class UIMgr
    {
        static private Transform mUIRootTransform;
        static public Transform UIRootTrsansform
        {
            get
            {
                if (mUIRootTransform == null)
                {
                    GameObject obj = GameObject.Find("UI Root");
                    if (obj != null)
                    {
                        mUIRootTransform = obj.transform;    
                    }
                }

                return mUIRootTransform;
            }
        }

        //ShowUI，UnShowUI方法接受的uiid，通过该字典查询对应的UIFrame
        //处理UI打开对于其他界面的互斥逻辑时，也需要对所有UI有一个统筹
        //所有该UIMgr里的方法，但凡传进来的参数是uiid，则都需要该字典
        static private Dictionary<int, UIFrame> mUIFrameDic = new Dictionary<int, UIFrame>();

        //UIFrame的UI状态事件方法
        static private LuaFunction mOnCreate;
        static private LuaFunction mOnEnable;
        static private LuaFunction mOnDisable;
        static private LuaFunction mOnDestroy;

        //NGUI事件方法
        static private LuaFunction mOnClick;
        static private LuaFunction mOnPress;
        static private LuaFunction mOnSelect;

        static private LuaFunction mOnDoubleClick;

        static private LuaFunction mOnDragStart;
        static private LuaFunction mOnDrag;
        static private LuaFunction mOnDragEnd;
        static private LuaFunction mOnDragOver;
        static private LuaFunction mOnDragOut;

        static public void Init(
            LuaFunction onCreate, LuaFunction onEnable, LuaFunction onDisable, LuaFunction onDestroy,
            LuaFunction onClick, LuaFunction onPress, LuaFunction onSelect,
            LuaFunction onDoubleClick,
            LuaFunction onDragStart, LuaFunction onDrag, LuaFunction onDragEnd, LuaFunction onDragOver, LuaFunction onDragOut)
        {
            mOnCreate = onCreate;
            mOnEnable = onEnable;
            mOnDisable = onDisable;
            mOnDestroy = onDestroy;

            mOnClick = onClick;
            mOnPress = onPress;
            mOnSelect = onSelect;

            mOnDoubleClick = onDoubleClick;

            mOnDragStart = onDragStart;
            mOnDrag = onDrag;
            mOnDragEnd = onDragEnd;
            mOnDragOver = onDragOver;
            mOnDragOut = onDragOut;
        }

        //下面两个方法调度UIFrame
        //UI关联的资源处理细节放到UIFrame里做
        static public void ShowUI(int uiID, string uiName)
        {
            Utils.Log("----------------------------ShowUI");
            UIFrame frame = null;
            if (!mUIFrameDic.TryGetValue(uiID, out frame))
            {
                frame = new UIFrame(uiID, uiName);
                mUIFrameDic[uiID] = frame;
            }

            frame.OnShow();
        }

        static public void UnShowUI(int uiID)
        {
            UIFrame frame = null;
            if (mUIFrameDic.TryGetValue(uiID, out frame))
            {
                frame.OnUnShow();
            }
        }

        //下面的方法是，UI的状态回调，由UIFrame发起
        //在这里做转发，传递给lua层
        //lua层再根据uiid派发给各个UI对应的lua脚本
        static public void OnCreate(UIFrame frame)
        {
            mOnCreate.BeginPCall();
            mOnCreate.Push(frame.UIID);
            mOnCreate.Push(frame);
            mOnCreate.PCall();
            mOnCreate.EndPCall();
        }

        static public void OnEnable(UIFrame frame)
        {
            mOnEnable.BeginPCall();
            mOnEnable.Push(frame.UIID);
            mOnEnable.Push(frame);
            mOnEnable.PCall();
            mOnEnable.EndPCall();
        }

        static public void OnDisable(UIFrame frame)
        {
            mOnDisable.BeginPCall();
            mOnDisable.Push(frame.UIID);
            mOnDisable.Push(frame);
            mOnDisable.PCall();
            mOnDisable.EndPCall();
        }

        static public void OnDestroy(UIFrame frame)
        {
            mOnDestroy.BeginPCall();
            mOnDestroy.Push(frame.UIID);
            mOnDestroy.Push(frame);
            mOnDestroy.PCall();
            mOnDestroy.EndPCall();
        }

        //下面的方法是各种UI事件方法，由UIEvent调用，直接传递到lua层
        //在UIMgr里接受lua注册进来的各种UI事件的方法，不同的界面以uiid来区分
        //在lua层根据uiid再做一次派发，直到每个UI事件传递到UI对应的lua脚本里
        static public void OnClick(int uiID, int eventID)
        {
            mOnClick.BeginPCall();
            mOnClick.Push(uiID);
            mOnClick.Push(eventID);
            mOnClick.PCall();
            mOnClick.EndPCall();
        }

        static public void OnPress(int uiID, bool isPressed, int eventID)
        {
            mOnPress.BeginPCall();
            mOnPress.Push(uiID);
            mOnPress.Push(isPressed);
            mOnPress.Push(eventID);
            mOnPress.PCall();
            mOnPress.EndPCall();
        }

        static public void OnSelect(int uiID, bool selected, int eventID)
        {
            mOnSelect.BeginPCall();
            mOnSelect.Push(uiID);
            mOnSelect.Push(selected);
            mOnSelect.Push(eventID);
            mOnSelect.PCall();
            mOnSelect.EndPCall();
        }

        static public void OnDoubleClick(int uiID, int eventID)
        {
            mOnDoubleClick.BeginPCall();
            mOnDoubleClick.Push(uiID);
            mOnDoubleClick.Push(eventID);
            mOnDoubleClick.PCall();
            mOnDoubleClick.EndPCall();
        }

        static public void OnDragStart(int uiID, int eventID)
        {
            mOnDragStart.BeginPCall();
            mOnDragStart.Push(uiID);
            mOnDragStart.Push(eventID);
            mOnDragStart.PCall();
            mOnDragStart.EndPCall();
        }

        static public void OnDrag(int uiID, Vector2 delta, int eventID)
        {
            mOnDrag.BeginPCall();
            mOnDrag.Push(uiID);
            mOnDrag.Push(delta);
            mOnDrag.Push(eventID);
            mOnDrag.PCall();
            mOnDrag.EndPCall();
        }

        static public void OnDragEnd(int uiID, int eventID)
        {
            mOnDragEnd.BeginPCall();
            mOnDragEnd.Push(uiID);
            mOnDragEnd.Push(eventID);
            mOnDragEnd.PCall();
            mOnDragEnd.EndPCall();
        }

        static public void OnDragOver(int uiID, GameObject obj, int eventID)
        {
            mOnDragOver.BeginPCall();
            mOnDragOver.Push(uiID);
            mOnDragOver.Push(obj);
            mOnDragOver.Push(eventID);
            mOnDragOver.PCall();
            mOnDragOver.EndPCall();
        }

        static public void OnDragOut(int uiID, GameObject obj, int eventID)
        {
            mOnDragOut.BeginPCall();
            mOnDragOut.Push(uiID);
            mOnDragOut.Push(obj);
            mOnDragOut.Push(eventID);
            mOnDragOut.PCall();
            mOnDragOut.EndPCall();
        }
    }

}